### ISSUE TITLE
string.buffer: gc not keeping up with allocations

### ISSUE THREAD

**Issue by capr** [LuaJIT/LuaJIT#872]
Looks like the gc doesn't know the amount of memory allocated by the buffer object so the buffer is not putting enough pressure on it to start a sweep.

This OOMs fast:

```lua
local buffer = require'string.buffer'
while 1 do
	buffer.new():reserve(65536)
	print(collectgarbage'count')
end
```

while this doesn't:

```lua
while 1 do
	local b = ffi.new('char[?]', 65536)
	print(collectgarbage'count')
end
```

This brings up something that I wanted to ask since forever: add an optional `size` arg to `ffi.gc()`. Without this arg, any externally allocated object cannot be properly tracked by Lua's gc, taking us back to manual freeing of such objects.

Great API btw, when will it be published? :)

**Comment by MikePall** [comment]
The lack of tracking for indirect memory use of userdata objects is a well-known weakness of Lua. But this is not easy to fix in general. At this time, I'm more inclined to add a workaround just for string buffers. I'm keeping this open, but don't wait for it.

I think the string buffer API is quite OK for now. Maybe we should start a thread on the mailing list to ask for further design input (without bike-shedding, e.g. put() vs. write()). I'll certainly add pattern matching (like string.match), but this is pending on a complete rewrite of the pattern matcher (I have ensured sponsorship and I have a design in the drawer, but lack the time).

**Comment by capr** [comment]
A workaround would be great. I actually thought of reimplementing the whole thing in Lua because of this issue but in the end I just added free() calls everywhere to plug the leaks.

AFAIC I like the API, it simplified my code a lot. I used it for consuming http and smtp streams which helped me get rid of a "line buffer" abstraction that everybody makes for line-based protocols so I was happy for that, and I also used it for a binary protocol called DIMSE, and also for a simple protocol for structured remote logging (since... free serialization). For all that it worked great, and got even better after I made it I/O-aware[1] to the point where it is now my API of choice for consuming files and sockets (which was made easier by the fact that the file/socket APIs themselves are cdata-based). One minor annoyance was the fact that the buffer object is a userdata instead of a table so I couldn't just monkey-patch it with new fields and methods like I'm used to, I had to wrap it and forward methods to it, no big deal.

Btw, is there a LuaJIT technical paper / book in your drawer as well ? :)

[1] https://github.com/allegory-software/allegory-sdk/blob/dev/lua/pbuffer.lua

**Comment by MikePall** [comment]
Some string buffer methods had missing GC checks. I've fixed this. Thanks!

But more importantly, the Lua 5.1 GC (on which the LuaJIT GC is based on) apparently made wrong estimates for managed memory which is attached to userdata. I've fixed that, too.

This works well for string.buffers, but not for userdata or cdata with attached unmanaged memory (malloc/free, mmap/munmap). Yes, I'm considering to add support for the cdata case in the future, but this is tricky and has to wait.
